packages/block-editor/src/hooks/color.js

```diff
@@ -0,0 +1,247 @@
+/**
+ * External dependencies
+ */
+import { assign, has } from 'lodash';
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { hasBlockSupport } from '@wordpress/blocks';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { __ } from '@wordpress/i18n';
+import { useState, useEffect } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+// This should be moved out of the components folder
+import {
+	getColorClassName,
+	getColorObjectByColorValue,
+	getColorObjectByAttributeValues,
+} from '../components/colors';
+import PanelColorSettings from '../components/panel-color-settings';
+import ContrastChecker from '../components/contrast-checker';
+import InspectorControls from '../components/inspector-controls';
+import { getBlockDOMNode } from '../utils/dom';
+
+export const COLOR_SUPPORT_KEY = '__experimentalColor';
+
+/**
+ * Filters registered block settings, extending attributes to include
+ * `backgroundColor` and `textColor` attribute.
```

Comment:
Should we try to have a "style" attribute that blocks and global styles can use to store global styles related information, instead of using the hook to register a new attribute?

Reply:
That's already the case in this PR. These extra attributes are needed for the color palette classNames unless there's a way to define these with global styles that I'm not aware of?

Reply:
I think @mtias suggested that instead of using classes for predefined colors we should still use CSS vars anyway but the CSS vars would reference another vars.

Reply:
I guess we may have a way to store things in style that indicates another variable is being referenced.
Global styles may have a mechanism where themes structurally set the variables, and we can say colors variables should reference some set, so when we set a color to a value equal to the variable part of a set instead of storing the value we reference the variable.

Reply:
Can I have an example of CSS for named colors using variables?

Reply:
So in theme.json a theme may provide something like:
```
colors: [
	{	name: "Red", slug: "red", color: "#f00" },
	{	name: "Blue", slug: "blue", color: "#00f" },
]
```
This would generate the following vars during the compilation:
```
--wp--colors--red: #f00;
--wp--colors--blue: #00f;
```

When storing a color attribute if the value is equal to one of the color variables instead of storing the value we would store a reference to the variable: `--wp--background-color: var(--wp--colors--red);`

In the style attribute, we may use the var directly `{"style":{"backgroundColor":"var(--wp--colors--red)"}}` or we may use a special reference mechanism syntax `{"style":{"backgroundColor":"ref:red"}}`

This solution has some disadvantages when compared with class usage:
- A theme can not automatically apply a text color when a specific background color is used, and not text color is selected.
- A theme can not apply a hover color depending on the text color.
- A theme can not define the colors of links in a block, depending on the text color selected.

Reply:
This is an interesting proposal and I can be onboard with that. Though, that's a bigger separate project. If we do this, we need to do it across blocks by measuring the pros and cons and communicating the changes properly.

For this PR, I think we can keep the existing behavior and it can still be adapted in the future if we use CSS variables for palettes too.

Reply:
If we would use CSS variables for registered theme colors, we could avoid having all those class names stored in HTML content generated by the `save` method altogether.

I think it feels like the very next step to explore. In theory, it could also open doors for having only one attribute used per color type as you could use CSS syntax for values.